auto-sync opcode & operand encoding info generation

AngelDev06 opened the pull request:
This is the modified llvm-capstone version that I used to generate the encoding info of both opcode and operands of an instruction for the **ARM** architecture. However this version could also be used to generate the required tables for the rest of the auto-sync archs. My modifications are quite simple. I only added two new functions, one called `getCSOperandEncoding` and the other `getCSOpcodeEncoding`. The former is being used on `printInsnOpMapEntry` (so the operand encoding info goes to the InsnMapOp.inc file) while the latter on `printInsnMapEntry`.

AngelDev06 committed:
Added operand and opcode info generation

XVilka commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -22,6 +22,7 @@
+#include <bitset>
 
 static void emitDefaultSourceFileHeader(raw_ostream &OS) {
   OS << "/* Capstone Disassembly Engine, http://www.capstone-engine.org */\n"
```
By the way, we should fix the URL(s) to be HTTPS

Rot127 replied:
```suggestion
  OS << "/* Capstone Disassembly Engine, https://www.capstone-engine.org */\n"
```

Rot127 replied:
Lets do it here. We will generate all inc files before the next release anyway.

Rot127 commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2524,6 +2490,180 @@ std::string getCSOperandType(Record const *OpRec) {
+  struct {
+    unsigned OperandPiecesCount;
+    std::array<unsigned, 5> Indexes;
+    std::array<unsigned, 5> Sizes;
```
In Capstone the indices and sizes array is of size 8. Please increase the size here and add a check below. If there are more pieces you can exit with something like `PrintFatalNote("Too many operands pieces etc.")` or `llvm_unreachable` or something.

AngelDev06 replied:
Originally I was writting the code with the mindset that it should only be used for ARM so I added 5 because that's actually the max number of operand pieces you can get. But anyway I fixed it.

Rot127 commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2524,6 +2490,180 @@ std::string getCSOperandType(Record const *OpRec) {
+  }
+
+  // if references were found, we add the encoding to the string
+  if (EncodingData.OperandPiecesCount) {
```
Please check for
```
if (!EncodingData.OperandPiecesCount)
   return "{ 0 }";
```

So the larger code block is not indented.

Rot127 commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2524,6 +2490,180 @@ std::string getCSOperandType(Record const *OpRec) {
+    if (auto const *const Bit =
+            dyn_cast<BitInit>(InstrBits->getBit(BitCount - InstrBitIdx - 1))) {
+      OpcodeData.Bits.set(OpcodeData.BitCount, Bit->getValue());
+      OpcodeData.Indexes[OpcodeData.BitCount] = InstrBitIdx - StartIdx;
```
Same as above. Please check for array overflows and fail safely. There is no target (to my knowledge) with >64bit instructions, but who knows what the future brings.

Rot127 reviewed (approved):
Please fix those comments. But besides that LGTM. Thanks a lot!

XVilka commented:
@AngelDev06 please rebase as AArch64 PR was merged and now there are some conflicts

AngelDev06 commented:
Done @XVilka

AngelDev06 committed:
Wrapped deprecated macro under an intellisense check

Basically intellisense fails, causing multiple errors in other files,

so when intellisense parses the code it will use the different version of the macro

wargio commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2537,7 +2537,193 @@ std::string getArchSupplInfo(StringRef const &TargetName,
+        if (OpRec->getValueAsDef("Type")->getValueAsInt("Size") == 0)
+            // Pseudo type
+            return "CS_OP_INVALID";
+        OperandType = "OPERAND_IMMEDIATE";
+    }
+    if (OperandType == "OPERAND_PCREL" || OperandType == "OPERAND_IMMEDIATE")
+        OperandType = "CS_OP_IMM";
```
why not just 
```suggestion
        return "CS_OP_IMM";
    }
    if (OperandType == "OPERAND_PCREL" || OperandType == "OPERAND_IMMEDIATE")
        OperandType = "CS_OP_IMM";
```

Rot127 commented on llvm/utils/TableGen/PrinterCapstone.cpp:
```diff
@@ -2537,7 +2537,193 @@ std::string getArchSupplInfo(StringRef const &TargetName,
   if (TargetName == "PPC")
     return getArchSupplInfoPPC(TargetName, CGI, PPCFormatEnum);
-  return "{{ 0 }}";
+  return "{ 0 }";
```
Please do not remove those brackets. They are necessary. Because without them we get a gazillion warnings from gcc because of this bug from the middle ages:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53119

AngelDev06 replied:
I don't recall removing them but anyway it's fixed now

AngelDev06 committed:
Fixed a small bug

Used double braces to prevent an old bug
